### PR TITLE
kube-prometheus: PVC (using persistent storage for the prom tsdb) example/documentation

### LINKED ISSUE #1819: What is the best way to specify storage with jsonnet setup?
Helm charts define `storageSpec` variable - which makes life super easy.

What is the best way to specify persistent storage with jsonnet setup?
If it can be provisioned before compilation with that `_config++`?
Or best way atm is to edit compiled yaml manifests and include PVCs volumeClaimTemplate there?

I'm talking about three main end-user components which might require persistent storage: prometheus, alertmanager, grafana.

[Storage](https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/storage.md) documentation seems to be a little bit blurry. Not enough examples.
> The StorageClass that was created can be specified in the storage section in the Prometheus resource.

What does it mean ? should i insert somewhere in jsonnet? But example below that lines looks like normal `yaml`.

It will be super helpful to see real examples. Like showing what and where to insert.

Just IMHO editing compiled yamls looks like easiest way atm, please correct me if I'm wrong :/

### PR DESCRIPTION
**Pull request by jolson490**
Resolves #1819

### CODE REVIEW

**Comment by jolson490** [comment]
Can someone (e.g. @brancz ) pls tell me how to get [jsonnet/prometheus-operator/prometheus-crd.libsonnet](https://github.com/coreos/prometheus-operator/blob/master/jsonnet/prometheus-operator/prometheus-crd.libsonnet) & [example/prometheus-operator-crd/prometheus.crd.yaml](https://github.com/coreos/prometheus-operator/blob/master/example/prometheus-operator-crd/prometheusrule.crd.yaml) & [contrib/kube-prometheus/manifests/0prometheus-operator-0prometheusCustomResourceDefinition.yaml](https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/manifests/0prometheus-operator-0prometheusCustomResourceDefinition.yaml) updated per the changes I made to `Documentation/api.md` in [this commit](https://github.com/coreos/prometheus-operator/pull/1837/commits/c8b6a30)? I ran `make generate-in-docker` in my directory where i've git cloned this repo and made my changes for this PR (and i made sure those changes are also present in `$GOPATH/src/github.com/coreos/prometheus-operator`), but running `make generate-in-docker` didn't update my local copy of any of the aforementioned 3 files - more info:
 * after running `make generate-in-docker` from the `prometheus-operator/` folder (per [this](https://github.com/coreos/prometheus-operator/blob/master/README.md#contributing)), I did `git status` I saw `modified:   Documentation/api.md` (it seems like `make generate-in-docker` undid my changes from my local copy of `api.md`).
 * running `make generate-in-docker` from the `prometheus-operator/contrib/kube-prometheus/` folder (per [this](https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/README.md#contributing)) didn't make any changes to any files in my local folder.

File: contrib/kube-prometheus/examples/prometheus-pvc.jsonnet [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+          // The possible values for a prometheus <duration> are:
+          //  * https://github.com/prometheus/common/blob/c7de230/model/time.go#L178 specifies "^([0-9]+)(y|w|d|h|m|s|ms)$" (years weeks days hours minutes seconds milliseconds)
+          retention: "30d",
+
+          // Reference info: https://github.com/coreos/prometheus-operator/blob/master/Documentation/user-guides/storage.md
+          // By default (if the following 'storage.volumeClaimTemplate' isn't created), prometheus will be created with an EmptyDir for the 'prometheus-k8s-db' volume (for the prom tsdb).
+          // This 'storage.volumeClaimTemplate' causes the following to be automatically created (via dynamic provisioning) for each prometheus pod:
+          //  * PersistentVolumeClaim (and a corresponding PersistentVolume)
+          //  * the actual volume (per the StorageClassName specified below)
+          storage: { // https://github.com/coreos/prometheus-operator/blob/master/Documentation/api.md#storagespec
```
Comment (jolson490) [inline contrib/kube-prometheus/examples/prometheus-pvc.jsonnet]: using this, when i deployed k-p to a K8s cluster in AWS, the Name of the AWS EBS Volume that was automatically created starts with "kubernetes-dynamic-pvc-" (and then has a bunch of random/generated letters & digits). Can anyone (perhaps @brancz ) tell me how to get a given specified string of text to be included somewhere in the EBS Volume? (The company I work for would like to be able to easily identify which of our teams are creating/using which cloud resources.)

More info:

I tried adding the following line...:
`pvc.mixin.metadata.withGenerateName('my-team-name'),`

...but the EBS that was created still had the same generic name format - and the only difference i noticed was the following (note that "generateName: my-team-name" is under the "metadata" - but "my-team-name" doesn't show up anywhere in the EBS Volume):
```
core@ip-192-168-xx-xx ~ $ kubectl -n monitoring get pvc prometheus-k8s-db-prometheus-k8s-0 -o yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/aws-ebs
  creationTimestamp: 2018-08-30T14:34:13Z
  finalizers:
  - kubernetes.io/pvc-protection
  generateName: my-team-name
  labels:
    app: prometheus
    prometheus: k8s
  name: prometheus-k8s-db-prometheus-k8s-0
  namespace: monitoring
  resourceVersion: "3637001"
  selfLink: /api/v1/namespaces/monitoring/persistentvolumeclaims/prometheus-k8s-db-prometheus-k8s-0
  uid: c3f40e8d-ac61-11e8-a38a-1288318eea14
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 100Gi
  storageClassName: ssd
  volumeName: pvc-c3f40e8d-ac61-11e8-a38a-1288318eea14
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 100Gi
  phase: Bound
```

(I Googled, & [this](https://github.com/ksonnet/parts/blob/master/incubator/redis/redis.libsonnet) is the only e.g. I could find of [ksonnet persistentVolumeClaim](http://g.bryan.dev.hepti.center/core/v1/persistentVolumeClaim/#core.v1.persistentVolumeClaim) being used.)

Comment (brancz) [reply contrib/kube-prometheus/examples/prometheus-pvc.jsonnet]: Through the PersistentVolume you can find which PersistentVolumeClaim has claimed it, that way you can make the correlation between namespace/object name and the team it belongs to.

**Comment by brancz** [comment]
It looks like generating worked successfully now. For future reference if it's not generating what you expect try `make clean && make generate-in-docker`.

**Comment by brancz** [comment]
This is a really awesome example. Thanks a lot for contributing this! :+1:

**Pull request merged**